### PR TITLE
Implicitly enable -g when using -C tracer=hw

### PR DESCRIPTION
**Pull request by ptersilie**
Since hardware tracing currently only runs when debugging is enabled, we implicitly enable `-g` automatically when setting `-C tracer=hw`.

### CODE REVIEW

**Comment by vext01** [comment]
Any idea what's going on here @ptersilie?

**Comment by ptersilie** [comment]
Not really. I ran this on bencher13 and it works fine.

**Comment by vext01** [comment]
I'm not sure either. You're going to have to investigate I'm afraid.

**Comment by ptersilie** [comment]
bors try

**Comment by ptersilie** [comment]
bors try

**Comment by vext01** [comment]
Yeah, so it failed again:

```
---- [ui] ui/abi/stack-probes-lto.rs stdout ----
error: test compilation failed although it shouldn't!
status: signal: 6
command: "/home/buildbot-worker8/worker8/ykrustc-buildbot-build-script-hwt/build/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/home/buildbot-worker8/worker8/ykrustc-buildbot-build-script-hwt/build/src/test/ui/abi/stack-probes-lto.rs" "-Zthreads=1" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "-Zui-testing" "-o" "/home/buildbot-worker8/worker8/ykrustc-buildbot-build-script-hwt/build/build/x86_64-unknown-linux-gnu/test/ui/abi/stack-probes-lto/a" "-Crpath" "-O" "-Cdebuginfo=0" "-Zunstable-options" "-Lnative=/home/buildbot-worker8/worker8/ykrustc-buildbot-build-script-hwt/build/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-C" "lto" "-L" "/home/buildbot-worker8/worker8/ykrustc-buildbot-build-script-hwt/build/build/x86_64-unknown-linux-gnu/test/ui/abi/stack-probes-lto/auxiliary"
stdout:
------------------------------------------
------------------------------------------
stderr:
------------------------------------------
rustc: /home/buildbot-worker8/worker8/ykrustc-buildbot-build-script-hwt/build/src/llvm-project/llvm/include/llvm/CodeGen/SlotIndexes.h:398: llvm::SlotIndex llvm::SlotIndexes::getInstructionIndex(const llvm::MachineInstr&) const: Assertion `!BundleNonDebug.isDebugInstr() && "Could not use a debug instruction to query mi2iMap."' failed.
------------------------------------------
```

Something to do with LLVM API.

**Comment by bjorn3** [comment]
> Assertion `!BundleNonDebug.isDebugInstr() && "Could not use a debug instruction to query mi2iMap."' failed.

Maybe some place for enabling debuginfo has been missed?

**Comment by ptersilie** [comment]
bors try

**Comment by ptersilie** [comment]
So, it seems like force-enabling debug info with hardware tracing isn't that simple. The tests trigger an assertion here: https://github.com/llvm-mirror/llvm/blob/master/include/llvm/CodeGen/SlotIndexes.h#L390

The code doesn't expect to see a debug instruction here, and thus breaks if we force-enable debug info. I'm not sure yet how we can get around this, but I'll keep looking.

Edit: It looks like `skipDebugInstructionsForward` is supposed to skip over any debug instructions, though, so I'm wondering if our inserted YK debug strings are in any way special and thus not skipped.

**Comment by vext01** [comment]
One option might be to unset `STD_TRACER_MODE` for the suites which it interferes with

**Comment by ptersilie** [comment]
I'm not sure if this won't bite us someday. From looking at the code, it should skip debug instructions anyways, so I'm just not sure why it doesn't seem to skip our debug labels. Unless these tests are really not meant to be run in debug mode that is, of course.

**Comment by vext01** [comment]
I'd recommend staring at the code a bit longer

**Comment by vext01** [comment]
What's the status of this one? Any joy?

**Comment by ptersilie** [comment]
bors try

**Comment by ptersilie** [comment]
bors try

**Comment by ptersilie** [comment]
Okay, this finally passes now. I will have to squash this into a single commit and update the commit message to include some details about the optimisations problem. So let me know when you are done with the review.

File: src/librustc/session/config.rs [hunk 1]
```diff
@@ -2350,6 +2348,15 @@ pub fn build_session_options_and_crate_config(
...
         }
     };
+
+    // Disable hardware tracing when optimisations are enabled, switching to software tracing
+    // instead.
+    if opt_level != OptLevel::No {
+       cg.tracer = TracerMode::Software;
+    }
+
+    let cg = cg;
```
Comment (vext01) [inline src/librustc/session/config.rs:2358]: what's the `let cg = cg` for?

Comment (ptersilie) [reply src/librustc/session/config.rs:2358]: It makes `cg` immutable. It is original rustc code that I had to move down a bit in order to be able to mutate `cg.tracer`.

Comment (vext01) [reply src/librustc/session/config.rs:2358]: Ah. Clever. Never seen that before.

**Comment by vext01** [comment]
Please squash.

**Comment by ptersilie** [comment]
Have a look if that commit message works for you. I'm starting to think the optimisation changes should have come before this PR, but we only noticed them because this PR enabled hardware tracing in tests that didn't use `-g` before. But maybe it's fine if both changes are in a single commit. I'll leave that up to you.

**Comment by vext01** [comment]
> Since hardware tracing currently only runs when debugging is enabled

How about:

Since hardware tracing currently requires debugging to be enabled...

**Comment by ptersilie** [comment]
Commit message updated.

**Pull request merged**